# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two)

In [2]:
lambda x: x + 2

In [3]:
print(lambda x: x + 2)

In [4]:
# IIFE = Immediately Invoked Function Expression
# Will execute a Lambda function immediately after creation
# Syntax: (lambda x,y,etc.: expression)(arg1, arg2, etc)

(lambda x: x + 2)(10)
(lambda x: x + 2)(15)


#### Saving to a Variable

In [5]:
named_lambda = lambda y: y + 2

print(named_lambda)

<function <lambda> at 0x000001D75782D790>


In [6]:
named_lambda(20)

22

#### Multiple Inputs

In [7]:
def multiply(x, y):
    return x * y

multiply(10, 4)

40

In [8]:
# As an IIFE
(lambda x, y, z: x*y*z)(15, 3, 2)

90

#### Passing a Function into a Function

In [9]:
def apply_func(f, name):
    """
    f -> Function to execute
    name -> Name to execute Function on
    """
    return f(name)

def get_first_letter(string):
    return string[0]

apply_func(get_first_letter, "Terry")

'T'

#### Passing a Lambda into a Function

In [10]:
# Call the apply_func function, pass in a lambda function as argument 'f', pass 'codingtemple' as the name argument

apply_func(lambda string: string[-1], "codingtemple")

'e'

In [11]:
apply_func(lambda string: string.upper(), "lower to upper")

'LOWER TO UPPER'

In [12]:
def string_to_upper(string):
    return string.upper()

apply_func(string_to_upper, 'lower')

'LOWER'

#### Returning a Function from a Function

In [13]:
# Regularly defined function
# Takin in a 'num' argument and return a function

def create_multiplier(num):
    # Define a locally scoped function that takes an 'x' argument
    def inner_func(x):
        # Return x * the outer scopes 'num' value
        return x * num
    
    # Return the reference to the function defined within this create_multiplier function
    return inner_func

five_times = create_multiplier(5)
print(five_times)
five_times



<function create_multiplier.<locals>.inner_func at 0x000001D75786A4C0>


<function __main__.create_multiplier.<locals>.inner_func(x)>

In [14]:
five_times(10)

50

#### Returning a Lambda from a Function

In [15]:
# Create a function that takes in a num

def new_multiplier(num):
    # Return the lambda function that takes in an x and will multiply x by the outer scope's num
    return lambda x: x * num

four = new_multiplier(4)
print(four)
four(10)

<function new_multiplier.<locals>.<lambda> at 0x000001D75786A940>


40

#### If Statements within Lambdas

In [16]:
def do_something(number):
    if number > 10:
        return number**2
    else:
        return number + 10
    
do_something(15)
    


225

In [17]:
# Ternary: value_if_true if condition else value_if_false

def do_something_again(num):
    return num ** 2 if num > 10 else num + 10

do_something_again(15)


225

In [18]:
(lambda num: num**2 if num > 10 else num + 10)(15)

225

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [19]:
(lambda num: num **3)(10)


        

1000

In [20]:
f = lambda num: num ** 3

print(f(10))

1000


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [21]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [22]:
# Funtion that we want to apply to each element in a list

def squared_minus_two(num):
    return num ** 2 - 2


numbers = [3, 4, 7, 2, 7, 4, 9, 10, 14, 4]

# Create an empty list
new_nums = []
# Loop over the numbers list
for num in numbers:
    # Apply the function to that element
    new_num = squared_minus_two(num)
    new_nums.append(new_num)
    
print(new_nums)

[7, 14, 47, 2, 47, 14, 79, 98, 194, 14]


In [23]:
# Map(function, *iterables -> list, tuple, string, set, dict, etc.)

list(map(squared_minus_two, numbers))

[7, 14, 47, 2, 47, 14, 79, 98, 194, 14]

In [24]:
my_map_obj = map(squared_minus_two, numbers)

In [25]:
next(my_map_obj)

7

In [26]:
def multiply(x, y):
    print(f"x={x}, y={y}")
    return x*y
          
x_values = [1, 2, 3, 4, 5, 6]
y_values = [10, 20, 30, 40, 50, 60]
          
list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
x=6, y=60


[10, 40, 90, 160, 250, 360]

#### Using Lambda's with Map

In [28]:
numbers_again= [5,7,2,3,8,5,12,19,4,2]

list(map(lambda num: num**2-2, numbers_again))

[23, 47, 2, 7, 62, 23, 142, 359, 14, 2]

#### OLD In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [29]:
digits = [9,6,8,12,2]

list(map(lambda dig: dig*2-1, digits))



[17, 11, 15, 23, 3]

#### NEW In-Class Exercise #2

Use the map function to double each number if it is even and havle it if it is odd


In [32]:
new_digits = range(1,11)

list(map(lambda new_dig: new_dig * 2 if new_dig%2==0 else new_dig / 2, new_digits))

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [38]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [213]:
names = ['bob', 'billy', 'alex', 'amanda', 'amy', 'barbary', 'catherine']

def starts_with_a(name):
    return name[0].upper() == 'A'

a_names_map = list(map(starts_with_a, names))
print(a_names_map)

[False, False, True, True, True, False, False]


In [44]:
# Syntax: filter(funtion to apply, iterable -> List, Tuple, Set, Dict, String, etc...)

a_names_filter = list(filter(starts_with_a, names))
print(a_names_filter)

['alex', 'amanda', 'amy']


In [47]:
def remainder(num):
    # Takes a number and return remainder after dividing by 3
    return num % 3

my_nums = [x for x in range(1,11)]
print(my_nums)
list(map(remainder, my_nums))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]

In [48]:
list(filter(remainder, my_nums))

[1, 2, 4, 5, 7, 8, 10]

#### Using Lambda's with Filter()

In [49]:
list(filter(lambda name: True if name[0].upper() == 'B' else False, names))

['bob', 'billy', 'barbary']

In [50]:
list(filter(lambda name: name[0].upper() == 'B', names))

['bob', 'billy', 'barbary']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [56]:
from statistics import mean

In [62]:
nums = [76, 100, 6, 23, 30, 81, 31, 87, 2, 93, 57, 64, 41, 85, 51, 10, 17, 79, 23, 22]

print(mean(nums))
list(filter(lambda num: num > mean(nums), nums))

48.9


[76, 100, 81, 87, 93, 57, 64, 85, 51, 79]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [70]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [72]:
def add_nums(x, y):
    print('='*20)
    print(f'x: {x}')
    print(f"y: {y}")
    print(f"return {x + y}")
    return x+y

result_add = reduce(add_nums, [1,2,3,4,5])
print(result_add)

x: 1
y: 2
return 3
x: 3
y: 3
return 6
x: 6
y: 4
return 10
x: 10
y: 5
return 15
15


In [73]:
total = 0
for num in [1,2,3,4,5]:
    print(f"Total: {total}, Num: {num}")
    total += num
print(total)

Total: 0, Num: 1
Total: 1, Num: 2
Total: 3, Num: 3
Total: 6, Num: 4
Total: 10, Num: 5
15


#### Using Lambda's with Reduce()

In [82]:
reduce(lambda x, y: x * y, [1,2,3,4,5])

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [87]:
import time

def capitalize(name):
    # takes in a name, upper case the first letter and lowercase the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    # take in a first name, capitalize it, take in a last nime, capitale it, and then
    # returns the concatinated string of first + ' ' + last
    time.sleep(5)
    return capitalize(first) + " " + capitalize(last)

def greet_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")

greet_person('tErRY', 'KUBalEwSki', 'MoNDay')

Hello Terry Kubalewski, it sure is beautiful on this Monday


#### Implementing a Base Case

In [89]:
def add_nums(num):
    # Set up a base case
    if num <=1:
        print(f"add_nums({num}) = {num}")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)
        
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [97]:
# n! - n * n-1!
# 5! = 5 * 4!
# 4! = 4 * 3!
# ....
# 1! = 1

def factorial(num):
    if num <= 1:
        return 1
    else: 
        print(f"factorial({num}) = {num} * factorial({num - 1 })")
        return num * factorial(num-1)
    
factorial(10)

factorial(10) = 10 * factorial(9)
factorial(9) = 9 * factorial(8)
factorial(8) = 8 * factorial(7)
factorial(7) = 7 * factorial(6)
factorial(6) = 6 * factorial(5)
factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)


3628800

In [ ]:
def hide_the_cache():
    cache= {}
    def factorial(num):
        if num in cache:
            return cache[num]
        elif num  <= 1:
            return num
        else:
            answer = num * factorial(num-1)
            cache[num] = answer
            return answer
    return factorial

factorial(10)

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [121]:
def subtract(num):
    if num <= 1:
        return (num)
    else:
        return num - subtract(num - 1)

subtract(8)

4

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [122]:
lengths = map(len, ['Brian', 'Stanton', 'Coding Temple'])

print(lengths)

In [123]:
for item in lengths:
    print(item)

5
7
13


In [124]:
for item in lengths:
    print(item)

In [125]:
help(iter)

Help on built-in function iter in module builtins:

iter(...)
    iter(iterable) -> iterator
    iter(callable, sentinel) -> iterator
    
    Get an iterator from an object.  In the first form, the argument must
    supply its own iterator, or be a sequence.
    In the second form, the callable is called until it returns the sentinel.



In [126]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [136]:
a_new_list = [4,5,6,7,8,9,10]
list_iter = iter(a_new_list)
print(list_iter)

In [138]:
next(list_iter)

5

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [140]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        return start
        start += step
        
my_range_iter(0,10)

0

In [142]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        yield start
        start += step
        
y = my_range_iter(0,10)
print(y)

<generator object my_range_iter at 0x000001D7597D5510>


In [153]:
next(y)

StopIteration: 

In [154]:
for x in my_range_iter(10,50,2):
    print(x)

10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
48


#### Infinite Generator

In [176]:
# bad, never create infinite loops
def infinite(x):
    while True:
        yield x
        x += 1
        
infinite_gen = infinite(1)

In [209]:
next(infinite_gen)

33

StopIteration: 

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [106]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

list(filter(lambda place: len(place) > 2, places))

['Argentina', 'San Diego', 'Boston', 'New York']

In [107]:
sorted(list(filter(lambda place: len(place) > 2, places)))

['Argentina', 'Boston', 'New York', 'San Diego']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key" or the sorted function on the key</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [280]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


def sort_by_last(author):
    author.sort(key=lambda name: name.lower().split()[-1])
    return author



print(sort_by_last(author))


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [260]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



In [212]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [214]:
print(names)

print(sorted(names, key=len))

['bob', 'billy', 'alex', 'amanda', 'amy', 'barbary', 'catherine']
['bob', 'amy', 'alex', 'billy', 'amanda', 'barbary', 'catherine']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

list(map)

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [229]:
def fib(num):
    return num

fib(6)

6

In [231]:
n = 5

for i in range(n):
    print(f"Iteration {n}: {fib(i)}")

Iteration 5: 0
Iteration 5: 1
Iteration 5: 2
Iteration 5: 3
Iteration 5: 4


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [244]:

def num_squared(num):
    while num > 0:
        yield num * num
        num -= 1

do_it = num_squared(10)

In [246]:
next(do_it)

81